### PR TITLE
Next step in extractor runtime

### PR DESCRIPTION
**Pull request by einarmo**
This time it is implemented to work end-to-end, there are a few things remaining that should be implemented before it is actually done, but this is one major step further along.

### CODE REVIEW

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 1]
```diff
@@ -32,6 +69,320 @@ CancellationTokenSource source
...
+                if (backoff == 0)
+                {
+                    _activeLogger.LogInformation("Restarting extractor");
+                    backoff += 1;
+                    continue;
+                }
+
+                // Exponential backoff, with a fairly high minimum. 5 * 2^4 = 80 seconds, which should
+                // be enough time to not be unnecessarily harsh on the source system.
+                var backoffTime = TimeSpan.FromSeconds(5 * Math.Pow(2, Math.Min(4, backoff)));
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:153]: There is a cap, `Math.Min(4, backoff)`

**Comment by einarmo** [comment]
> Can we add more tests, preferably ones outside of the happy path?

Can this wait a bit, with a promise to add more before we actually finish this? Properly testing the unhappy path is going to be way more than 500 lines on its own, and we're already at the limit.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 2]
```diff
@@ -32,6 +69,320 @@ CancellationTokenSource source
...
+                        metrics.Start();
+                    }
+                    if (_params.AddLogger)
+                    {
+                        // Set the persistent active logger. This is somewhat of a hack, since it effectively
+                        // means we will only use the startup logger until we actually have a valid config,
+                        // but since the startup logger is often quite useless, this is better than nothing.
+                        // In the future we may want to add an option to _also_ log to the startup logger.
+                        // This might be handy for logging extractor events to e.g. windows event log.
+                        _activeLogger = provider.GetRequiredService<ILogger<TExtractor>>();
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:316]: Yuup. See the comment on this line. I do think we need a better way to deal with this, as discussed, but that can wait to a separate PR, I think. I'll add a TODO.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 3]
```diff
@@ -32,6 +69,320 @@ CancellationTokenSource source
...
+                    extractor = provider.GetRequiredService<TExtractor>();
+                    if (_params.OnCreateExtractor != null)
+                    {
+                        var destination = provider.GetService<CogniteDestination>();
+                        _params.OnCreateExtractor(destination, extractor);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    _activeLogger.LogError(ex, "Failed to build extractor: {msg}", ex.Message);
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:327]: This logs the exception, the first argument is the exception.

**Comment by ozangoktan** [comment]
> > Can we add more tests, preferably ones outside of the happy path?
> 
> Can this wait a bit, with a promise to add more before we actually finish this? Properly testing the unhappy path is going to be way more than 500 lines on its own, and we're already at the limit.

I would say not adding tests for them now would be a disservice to ourselves and the next reviewer. If it goes over the limit, it can't be helped. It is a big feature, and with how many lines the formatting uses, we are not that close to 500 lines of logic.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 4]
```diff
@@ -32,6 +69,320 @@ CancellationTokenSource source
...
+                    // Extractors should strive to report all possible config errors before
+                    // constructing the extractor itself.
+                    return ExtractorRunResult.EarlyError;
+                }
+
+                try
+                {
+                    // Do not wait for the cancellation token here, since we want to give the extractor the opportunity
+                    // to shut down cleanly.
+                    var waitTask = CommonUtils.WaitAsync(_revisionChangedEvent, Timeout.InfiniteTimeSpan, CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:338]: So, I get this, but I'm not sure we do. If the extractor fails to shut down, which would be a bug, then _what_ should we do? Crash the entire application? Just ignore that we now have _two_ instances of the extractor running? I honestly don't think there is a good solution here. A deadlock is in that case preferable. We also have no concept of how long it will take the extractor to shut down. It can be reasonable, that in the case of a huge amount of flushing being needed, it takes a few minutes to shut down the extractor. This is a framework, so we can't really make assumptions...

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:317]: We do have the _startupLogger property, why not make extensive use of that everywhere we don't have the configured logger active and double log once we know that it is configured?

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:328]: Ah, missed that.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 5]
```diff
@@ -32,6 +69,320 @@ CancellationTokenSource source
...
+                    }
+                    // Otherwise, immediately restart.
+                    _isFatal = false;
+                    backoff = 0;
+                }
+
+                if (backoff == 0)
+                {
+                    _activeLogger.LogInformation("Restarting extractor");
+                    backoff += 1;
```
[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:147]: But the only way for a tight restart loop to happen is with repeated ExtractorRunResult.CleanShutdown without the restart policy block taking effect, and this backoff counter does nothing to prevent that given we are doing continue right below. So this backoff number effectively does nothing except add +1 backoff to the other cases in the enum ExtractorRunResult if they are following an ExtractorRunResult.CleanShutdown.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:153]: Ah, you are right.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:317]: The problem is that the two loggers might be the _same_. If the startup logger is a console logger, then we will log double. This is why this is a much harder problem to solve than just "log to both always".

[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:317]: Hmm, I see, but can't we set the _activeLogger back to the startup logger once a crash happens?

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ozangoktan) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:339]: I am ok with that.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:316]: Can we deal with this later? I really think a patch here serves no real purpose if we're going to do something way more complicated later anyway.

**Comment by einarmo** [comment]
I've added tests. Those also required some changes, and uncovered some edge cases.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 6]
```diff
@@ -32,6 +70,326 @@ CancellationTokenSource source
...
+                // Reset the revision changed event as late as possible, to avoid
+                // restarting unnecessarily.
+                _revisionChangedEvent.Reset();
+
+                var newConfig = await _configSource.ResolveConfig(null, bootstrapErrorReporter, _source.Token).ConfigureAwait(false);
+                if (_isFatal && !newConfig)
+                {
+                    _activeLogger.LogDebug("No new config after fatal error, retrying");
+                    return ExtractorRunResult.EarlyError;
+                }
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:221]: Possibly... This is returned if the _extractor_ throws an error during setup. Since the extractor is user-provided code, it can in theory be caused by anything. Maybe we should have another error variant, like be explicit about `ConfigError`? I can try making that change.

[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:221]: I dislike it a bit, because it means that the extractor will try to start even with a config error, which may be noisy, but it's probably too risky.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 7]
```diff
@@ -32,6 +70,326 @@ CancellationTokenSource source
...
+                    // it, since this can really hit source systems hard.
+                    if (elapsed >= TimeSpan.FromSeconds(600))
+                    {
+                        backoff = 0;
+                    }
+                }
+                else if (result == ExtractorRunResult.CleanShutdown)
+                {
+                    // Shut down, if the extractor is configured to only restart on error.
+                    if (_params.RestartPolicy == ExtractorRestartPolicy.OnError)
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:135]: If it's never we already exited the function earlier, on line 112, but sure I can change it here as well. Probably more defensive.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs [hunk 8]
```diff
@@ -32,6 +70,326 @@ CancellationTokenSource source
...
+            }
+
+            return exception;
+        }
+
+        /// <summary>
+        /// Run the extractor once.
+        /// </summary>
+        /// <returns>Returns the type of extractor termination.</returns>
+        private async Task<ExtractorRunResult> RunExtractorIteration()
```
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:199]: I'll try. Semantically it _is_ a single long function, with several early returns that aren't easily factored out in a nice way, but I can try.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (frxstrem) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs]: :+1:

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (frxstrem) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:204]: I was thinking more about pulling out chunks of code with a single purpose, not splitting into functions that chain. The scope of the function seems fine, it's more about the lack of readability for very long functions, so having the function have the same high level structure but replacing larger chunks of single-purpose code with function calls with clear names.

For example, this block of code seemed like a good candidate, which could instead have been one or a few function calls to extracted functions:
https://github.com/cognitedata/dotnet-extractor-utils/blob/27f1612f001b1f8de081a74feece5a89005d9ffb/ExtractorUtils/Unstable/Runtime/Runtime.cs#L232-L287

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (frxstrem) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs]: If you want to keep it as-is and the behavior is sensible for that case, then I think that's fine too.

File: ExtractorUtils/Unstable/Runtime/Runtime.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (einarmo) [reply ExtractorUtils/Unstable/Runtime/Runtime.cs:204]: Yeah, that bit I managed to split out. The problem is that there's no clean way to split out a part of the top-level function that _may_ return early.

I managed to split off the `service` bit, and I managed to create a separate function for the build-and-run stage. I don't see any good candidates for splitting further.

**Review by frxstrem — COMMENTED** [review]
🦄
